lib/child_process.js

```diff
@@ -250,45 +245,43 @@ exports.execFile = function(file /*, args, options, callback*/) {
   }
 
   if (child.stdout) {
-    stdoutState = child.stdout._readableState;
+    if (encoding)
+      child.stdout.setEncoding(encoding);
 
     child.stdout.addListener('data', function(chunk) {
-      // If `child.stdout.setEncoding()` happened in userland, convert string to
-      // Buffer.
-      if (stdoutState.decoder) {
-        chunk = Buffer.from(chunk, stdoutState.decoder.encoding);
-      }
-
-      stdoutLen += chunk.byteLength;
+      stdoutLen += chunk instanceof Buffer ?
+        chunk.length :
+        Buffer.byteLength(chunk, encoding);
 
       if (stdoutLen > options.maxBuffer) {
         ex = new Error('stdout maxBuffer exceeded');
-        stdoutLen -= chunk.byteLength;
         kill();
       } else {
-        _stdout.push(chunk);
+        if (!encoding)
+          _stdout.push(chunk);
+        else
+          _stdout += chunk;
       }
     });
   }
 
   if (child.stderr) {
-    stderrState = child.stderr._readableState;
+    if (encoding)
+      child.stderr.setEncoding(encoding);
 
     child.stderr.addListener('data', function(chunk) {
-      // If `child.stderr.setEncoding()` happened in userland, convert string to
-      // Buffer.
-      if (stderrState.decoder) {
-        chunk = Buffer.from(chunk, stderrState.decoder.encoding);
-      }
-
-      stderrLen += chunk.byteLength;
+      stderrLen += chunk instanceof Buffer ?
```

Comment:
Benchmark changing it to `encoding`:

```
child_process/child-process-exec-stdout.js len="64" dur="5": 40832.10138
child_process/child-process-exec-stdout.js len="256" dur="5": 40848.40837
child_process/child-process-exec-stdout.js len="1024" dur="5": 40868.37676
child_process/child-process-exec-stdout.js len="4096" dur="5": 40861.78442
child_process/child-process-exec-stdout.js len="32768" dur="5": 40869.98888
```

Benchmark leaving it as `instanceof Buffer`:

```
child_process/child-process-exec-stdout.js len="64" dur="5": 40864.86590
child_process/child-process-exec-stdout.js len="256" dur="5": 40855.07753
child_process/child-process-exec-stdout.js len="1024" dur="5": 40905.43913
child_process/child-process-exec-stdout.js len="4096" dur="5": 40882.65445
child_process/child-process-exec-stdout.js len="32768" dur="5": 40868.74424
```

I think `instanceof Buffer` makes it clearer what we're trying to test but agree that `encoding` *should* be more performant, even though it doesn't seem to be measurable, at least in our current benchmark.